domain/src/main/java/org/oppia/domain/oppialogger/crashlytics/ExceptionsController.kt

```diff
@@ -0,0 +1,174 @@
+      Throwable(exceptionLog.cause.message)
+    } else {
+      Throwable()
+    }
+    exceptionCause.stackTrace = createErrorStackTrace(exceptionLog.cause)
+    val exception = Exception(exceptionMessage, exceptionCause)
+    exception.stackTrace = createErrorStackTrace(exceptionLog)
+    return exception
+  }
+
+  /** Builds the [ExceptionLog.StackTraceElement] for an [exception] with relevant data and adds it to the [exceptionLogBuilder]. */
+  private fun buildStackTraceElement(
+    stackTraceSize: Int,
+    exceptionLogBuilder: ExceptionLog.Builder,
+    exception: Exception
+  ) {
+    for (i in 0 until stackTraceSize) {
+      val stackTraceElement = ExceptionLog.StackTraceElement.newBuilder()
+      stackTraceElement.fileName = exception.stackTrace[i].fileName
+      stackTraceElement.methodName = exception.stackTrace[i].methodName
+      stackTraceElement.lineNumber = exception.stackTrace[i].lineNumber
+      stackTraceElement.declaringClass = exception.stackTrace[i].className
+      exceptionLogBuilder.addStacktrace(i, stackTraceElement.build())
+    }
+  }
+
+  /** Returns exception stacktrace for the [exceptionLog]. */
+  private fun createErrorStackTrace(exceptionLog: ExceptionLog): Array<StackTraceElement> {
+    return Array(
+      exceptionLog.stacktraceCount,
+      init = { i: Int ->
+        StackTraceElement(
+          exceptionLog.stacktraceList[i].declaringClass,
+          exceptionLog.stacktraceList[i].methodName,
+          exceptionLog.stacktraceList[i].fileName,
+          exceptionLog.stacktraceList[i].lineNumber
+        )
+      }
+    )
+  }
+
+  /**
+   * Adds an exception to the storage.
+   *
+   * At first, it checks if the size of the store isn't exceeding [exceptionLogStorageCacheSize].
+   * If the limit is exceeded then the least recent exception is removed from the [exceptionLogStore].
+   * After this, the [exceptionLog] is added to the store.
+   * */
+  private fun cacheExceptionLog(exceptionLog: ExceptionLog) {
+    exceptionLogStore.storeDataAsync(true) { oppiaExceptionLogs ->
+      val storeSize = oppiaExceptionLogs.exceptionLogList.size
+      if (storeSize + 1 > exceptionLogStorageCacheSize) {
+        val exceptionLogRemovalIndex = getLeastRecentExceptionIndex(oppiaExceptionLogs)
+        if (exceptionLogRemovalIndex != null) {
+          return@storeDataAsync oppiaExceptionLogs.toBuilder()
+            .removeExceptionLog(exceptionLogRemovalIndex)
+            .addExceptionLog(exceptionLog)
+            .build()
+        } else {
+          // TODO (#1433): Refactoring for logging exceptions to both console and exception loggers.
+          val exception =
+            NullPointerException(
+              "Least Recent Exception index absent -- ExceptionLogCacheStoreSize is 0"
+            )
+          consoleLogger.e("Exceptions Controller", exception.toString())
+          logException(exception, oppiaClock.getCurrentCalendar().timeInMillis)
```

Comment:
I'm wondering if it makes sense to log an exception here if the problem lies in logging exceptions

Reply:
This might cause a crash/fail loop, so we probably shouldn't try this and instead fail in the console logger.

Reply:
Removed.